`Development:` Add a description of the new Artemis development process (#8089)
ls1intum/Artemis: docs/dev/development-process/development-process.rst

*******************
Development Process
*******************

.. contents:: Content of this document
    :local:
    :depth: 2

The Artemis Development Process Model
===========================================

.. figure:: ./artemis-process-model.png
    :align: center
    :alt: Artemis Process Model - Activity Diagram
    :figclass: align-center

1. Submit a Feature Request
---------------------------
The initial step in our development process involves the creation of a feature request, which is accomplished through the submission of a GitHub Issue. 
This action can be performed by any stakeholder, including developers, users, or maintainers. 
The feature request should include a detailed description of the desired functionality, as well as any relevant information that may be useful to the development team. 

> This information may include the rationale for the feature, the expected benefits, and any potential risks or challenges that may be associated with the implementation of the feature.

Reviewer: > may include

Using "may" in formal documents is considered poor style because its meaning is ambiguous. I recommend using a stronger word here, such as "should".

.. figure:: ./aremtis-feature-proposal-flow.png
    :align: center
    :alt: Create a Feature Proposal - Activity Diagram
    :figclass: align-center

Step 1: Amend Feature Proposal Template to Feature Request on GitHub
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The Feature Proposal Template outlines the structure of the feature proposal and provides a guideline for the development team to follow.

We have written a Github Action that automatically attaches the feature proposal template to the issue description of an issue when someone is assigned.
As a developer who is responsible for the implementation of a feature request, you do not have to do any than register as an assignee in the issue. 

.. note::
      The GitHub Action will be skipped for issues that are labeled with "bug".

.. literalinclude:: ../../../.github/ISSUE_TEMPLATE/feature-proposal--developer-.md
    :caption: Artemis/.github/ISSUE_TEMPLATE/feature-proposal--developer-.md
    :language: markdown

Step 2: Requirements Engineering
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> This section of the feature proposal includes a description of all Functional and Non-Functional Requirements that the feature should fulfill.

Reviewer: > Functional and Non-Functional Requirements

This doesn't need to be capitalized, they are important concepts but not proper nouns

Refer to the template for more details on the format. 

Step 3: Analysis
^^^^^^^^^^^^^^^^

> This section should outline the involved Analysis Objects (Analysis Object Model) and the dynamic behavior of the feature (Activity Diagram, State Chart diagram, Communication Diagram).

Reviewer: > Analysis Objects (Analysis Object Model)

"analysis objects" should be left lowercase, while AOM is a proper noun and should be capitalized

> This section should outline the involved Analysis Objects (Analysis Object Model) and the dynamic behavior of the feature (Activity Diagram, State Chart diagram, Communication Diagram).

Reviewer: Also I would capitalize the D in "State Chart Diagram"

Step 4: System Architecture
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> This step should include all involved subsystems and their interfaces (Subsystem Decomposition), as well as a description of the persistent data management (Data Model), access control / security asepcts and further architectural decisions.

Reviewer: > security asepcts

typo

1. **Initiate Design Process**: Upon receiving the UICompanion comment, click on the provided Figma Project Link to access our Figma project.
2. **Follow Guidance**: Inside Figma, use the UICompanion Figma app for step-by-step guidance through the design process. This includes creating prototypes and seeking approval from designated Artemis maintainers.
3. **Automate Tasks**: UICompanion helps automate tasks such as adding the necessary UI components from the Artemis Design System, ensuring that designs remain consistent with the project's UI guidelines.

For more details and to get started with UICompanion, visit the `UICompanion Repository on GitHub <https://github.com/ls1intum/UICompanion>`_.

4. Implement the Feature
------------------------
In this step, the development team will implement the feature according to the feature proposal.

5. Create a Pull Request
------------------------
Once the feature has been implemented, the developer will create a pull request to merge the feature into the develop branch.

Naming Conventions for GitHub Pull Requests
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> 1. The first term is a main feature of Artemis and is using code highlighting, e.g.  “``Programming exercises``:”.

Reviewer: > is using

"should use"

**2. Implementation is "done" -> only Developer**

* Make sure all steps in the `Checklist <https://github.com/ls1intum/Artemis/blob/develop/.github/PULL_REQUEST_TEMPLATE.md>`_ are completed.
* Add or update the "Steps for Testing" in the description of your pull request.
* Make sure that the changes in the pull request are only the ones necessary.
* Mark the pull request as `ready for review. <https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request>`_

**3. Review**

Developer:

* Organize or join a testing session. Especially for large pull requests this makes testing a lot easier.
* Actively look for reviews. Do not just open the pull request and wait.

Reviewer:

* Perform the "Steps for Testing" and verify that the new functionality is working as expected.
* Verify that related functionality is still working as expected.
* Check the changes to
    * conform with the code style.
    * make sure you can easily understand the code.
    * make sure that (extensive) comments are present where deemed necessary.

>     * performance is reasonable (e.g. number of database queries or HTTP calls).

Reviewer: > * Check the changes to
    * conform with the code style.
    * make sure you can easily understand the code.
    * make sure that (extensive) comments are present where deemed necessary.
    * performance is reasonable (e.g. number of database queries or HTTP calls).

The wording doesn't really work here. I would suggest:
 * Ensure that the code changes...
    * conform to the code style guide.
    * are easily understandable.
    * contain thorough documentation where applicable.
    * maintain reasonable performance (e.g. no excessive/unnecessary database queries or HTTP calls).

* Submit your comments and status (👍 Approve or 👎 Request Changes) using GitHub.
    * Explain what you did (test, review code) and on which test server in the review comment.

**4. Respond to review**

Developer:

* Use the pull request to discuss comments or ask questions.
* Update your code where necessary.
* Revert to draft if the changes will take a while during which review is not needed/possible.
* Set back to ready for review afterwards.
* Notify the reviewer(s) once your revised version is ready for the next review.
* Comment on "inline comments" (e.g. "Done").

Reviewer:

> * Respond to questions raised by the reviewer.

Reviewer: > reviewer

developer

* Mark conversations as resolved if the change is sufficient.

.. note::

>       Iterate steps 3 & 4 until ready for merge (all reviewers approve 👍)

Reviewer: > Iterate

Repeat

> ready for merge

ready to merge

>       Iterate steps 3 & 4 until ready for merge (all reviewers approve 👍)

Reviewer: Would also specify a target number of reviewers here (~4)

**5. Merge**

A project maintainer merges your changes into the ``develop`` branch.

Stale Bot
^^^^^^^^^

> If the pull request doesn't have any activity for at least 7 days, the stale bot will mark the PR as `stale`.

Reviewer: > stale bot

Proper noun, capitalize here

The `stale` status can simply be removed by adding a comment or a commit to the PR.

> After the PR is marked as `stale`, the bot waits another 14 days until the PR will be closed (21 days in total).

Reviewer: This should be rephrased.
"After 21 total days of inactivity and 14 days after marking the PR as stale, the Stale Bot will close the PR."

Adding activity to the PR will remove the `stale` label again and reset the stale timer.
To prevent the bot from adding the `stale` label to the PR in the first place, the `no-stale` label can be used.
This label should only be utilized if the PR is blocked by another PR or the PR needs help from another developer.

> A full documentation on this bit can be found here:

Reviewer: > A full documentation on this bit can be found here:

"Further documentation on the Stale Bot can be found here:"

===========================
The initial step in our development process involves the creation of a feature request, which is accomplished through the submission of a GitHub Issue.
This action can be performed by any stakeholder, including developers, users, or maintainers.
The feature request should include a detailed description of the desired functionality, as well as any relevant information that may be useful to the development team.
This information should include the rationale for the feature, the expected benefits, and any potential risks or challenges that may be associated with the implementation of the feature.

2. Evaluate Feature Request
===========================
Once a feature request has been submitted, the maintainers will evaluate the request together with the development team to determine its feasibility and potential impact on the system.

3. Create a Feature Proposal
===============================
If the feature request is deemed feasible, the development team will create a feature proposal that extensively describes the proposed feature. This step will consist of the artifacts mentioned in the model "Create Feature Proposal" below:

> .. figure:: ./aremtis-feature-proposal-flow.png

Reviewer 2: (Here as well)

In the Feature Proposal Template, you can find examples for structuring such functional and non-functional requirements

Step 3: Analysis
----------------
Within the analysis section, the emphasis is on understanding and documenting the feature's intended behavior and interactions within the system.
This includes creating an Analysis Object Model, which represents the static structure of the system using class diagrams to identify key entities and their relationships.
Additionally, dynamic aspects of the feature are explored through Activity Diagrams (for workflow representation), State Chart Diagrams (to model the feature's states and transitions), and Communication Diagrams (illustrating how objects interact to achieve the feature's functionality).
These models serve as a blueprint for the development and ensure a shared understanding among all people working on Artemis.

Step 4: System Architecture
---------------------------
In this step, the high-level structure of the system supporting the feature is outlined, focusing on analyzing the system components relevant to the implementation of the feature,

> subsystem decomposition, interfaces, persistent data management and security aspects.

Reviewer 3: We tend to use the serial comma

```suggestion
subsystem decomposition, interfaces, persistent data management, and security aspects.
```

The subsystem decomposition process is used to identify the main components of the system and define their responsibilities, which facilitates modular development and maintenance.
It is important to define the interfaces between these subsystems to ensure seamless interaction and integration.

> In addition, the development of a comprehensive data model is important for a clear understanding of how data is securely stored, retrieved and managed within the system, taking into account both logical and physical data perspectives.

Reviewer 3: ```suggestion
In addition, the development of a comprehensive data model is important for a clear understanding of how data is securely stored, retrieved, and managed within the system, taking into account both logical and physical data perspectives.
```

Security aspects, especially access control mechanisms, are considered to ensure data integrity and privacy.

Step 5: UI/UX Design
--------------------
If the proposed feature requires a change to the software's user interface, this section should include a detailed description of the proposed changes, as well as a mockup that illustrates the new user interface.

> The mockup should be created using Figma in conjunction with the Artemis Design System and should be added to the feature proposal as a screenshot.

Reviewer 3: I think that we can directly link the Figma Artemis Project here

```suggestion
The mockup should be created using [Figma in conjunction with the Artemis Design System](https://www.figma.com/files/team/1238833580691926997/Artemi) and should be added to the feature proposal as a screenshot.
```

> The mockup should be created using Figma in conjunction with the [Artemis Design System](https://www.figma.com/files/team/1238833580691926997/Artemis) and should be added to the feature proposal as a screenshot.

Reviewer 4: I suspect this link is in the wrong format?
<img width="990" alt="Screenshot 2024-05-08 at 11 30 45" src="https://github.com/ls1intum/Artemis/assets/13572283/a331989e-4479-4e4d-a658-1f1275b8aa12">

> Furthermore it is important to include a description of the user flow that references the dynamic model created in the analysis section.

Reviewer 3: ```suggestion
Furthermore, it is important to include a description of the user flow that references the dynamic model created in the analysis section.
```

.. figure:: ./uiux_workflow.png
    :align: center
    :alt: UI/UX Design Workflow - Activity Diagram
    :figclass: align-center

.. raw:: html

    <div style="text-align: center;">
        <iframe src="https://live.rbg.tum.de/w/artemisintro/46233?video_only=1&t=0" allowfullscreen="1" frameborder="0" width="600" height="300">
            Watch this video on TUM-Live.
        </iframe>
    </div>

4. Implement the Feature
========================
In this step, the development team converts the detailed plans and designs outlined in the functional proposal into working code. 

> This step requires careful adherence to the previously defined requirements and system architecture to ensure that the function fits seamlessly into the existing system and fulfills the specified functional and performance criteria.

Author: Yes, developers should update the feature proposal to reflect any deviations from the initially defined system architecture due to unforeseen circumstances during the implementation phase. This ensures that the proposal represents the implemented state. Whether re-approval from the maintainers is necessary should be determined on a case-by-case basis in consultation with the respective maintainer.

.. note::
      Make sure to follow the `Artemis Code and Design Guidelines <https://docs.artemis.cit.tum.de/dev/guidelines.html>`_.

5. Create a Pull Request
========================
After the feature implementation is complete, the developer is required to create a pull request for integrating the feature into the develop branch. 
The subsequent sections provide guidance on the naming conventions and outline the necessary steps for creating and merging a pull request.

Naming Conventions for GitHub Pull Requests
-------------------------------------------

1. The first term is a main feature of Artemis and should use code highlighting, e.g.  ``Programming exercises``:

>     1. Possible feature tags are: ``Programming Exercises``, ``Integrated Code Lifecycle``, ``Quiz Exercises``, ``Modeling Exercises``, ``Text Exercises``, ``File Upload Exercises``, ``Exam Mode``,
>        ``Grading``, ``Assessment``, ``Communication``, ``Notifications``, ``Team Exercises``, ``Lectures``, ``Integrated Markdown Editor``, ``Plagiarism Checks``, ``Learning Analytics``,
>        ``Adaptive Learning``, ``Learning Path``, ``Tutorial Groups``, ``Iris``, ``Scalability``, ``Usability``, ``Performance``, ``Intrastructure``, ``Mobile Apps``.

Reviewer 5: Please revert the changes to the capitalization, as we do not capitalize terms after the first word (even if this is not according to english title capitalization)
```suggestion
    1. Possible feature tags are: ``Programming exercises``, ``Integrated code lifecycle``, ``Quiz exercises``, ``Modeling exercises``, ``Text exercises``, ``File upload exercises``, ``Exam mode``,
       ``Grading``, ``Assessment``, ``Communication``, ``Notifications``, ``Team exercises``, ``Lectures``, ``Integrated markdown editor``, ``Plagiarism checks``, ``Learning analytics``,
       ``Adaptive learning``, ``Learning path``, ``Tutorial groups``, ``Iris``, ``Scalability``, ``Usability``, ``Performance``, ``Intrastructure``, ``Mobile apps``.
```

Reviewer 2: Can you double-check the capitalization as mentioned by @rstief?

3. Review Process
^^^^^^^^^^^^^^^^^
**For Developers:**

* Organize or join a testing session. Especially for large pull requests this makes testing a lot easier.
* Actively look for reviews. Do not just open the pull request and wait.

**For Reviewers:**

* Perform the "Steps for Testing" and verify that the new functionality is working as expected.
* Verify that related functionality is still working as expected.
* Ensure that the code changes...
    * conform to the code style guide.
    * are easily understandable.
    * contain thorough documentation where applicable.
    * maintain reasonable performance (e.g. no excessive/unnecessary database queries or HTTP calls).
* Submit your comments and status (✅ Approve or ❌ Request Changes) using GitHub.
    * Explain what you did (test, review code) and on which test server in the review comment.

4. Address Review Comments
^^^^^^^^^^^^^^^^^^^^^^^^^^
**For Developers:**

* Use the pull request to discuss comments or ask questions.
* Update your code where necessary.
* Revert to draft if the changes will take a while during which review is not needed/possible.
* Set back to ready for review afterwards.

> * After making revisions, re-request reviews through GitHub initially. If no response is received, then follow up with a targeted message on Slack to the relevant reviewers.

Reviewer 2: ```suggestion
* After making revisions, re-request reviews through GitHub initially. If you receive no response, then follow up with a targeted message on Slack to the relevant reviewers.
```

*******************
Development Process
*******************

.. figure:: ./artemis-process-model.png
    :align: center
    :alt: Artemis Process Model - Activity Diagram
    :figclass: align-center

1. Submit a Feature Request
===========================
The initial step in our development process involves the creation of a feature request, which is accomplished through the submission of a GitHub Issue. 
This action can be performed by any stakeholder, including developers, users, or maintainers. 
The feature request should include a detailed description of the desired functionality, as well as any relevant information that may be useful to the development team. 

> This information should include the rationale for the feature, the expected benefits, and any potential risks or challenges that may be associated with the implementation of the feature.

Author: Again, the process documented here reflects what was discussed and agreed upon in multiple developer meetings. This current documentation simply implements the previously established process and is not set in stone. If you have concerns regarding this process, please bring them up at the developer meeting.

2. Evaluate Feature Request
===========================

> Once a feature request has been submitted, the maintainers will evaluate the request together with the development team to determine its feasibility and potential impact on the system.

Author: Again, the process documented here reflects what was discussed and agreed upon in multiple developer meetings. This current documentation simply implements the previously established process and is not set in stone. If you have concerns regarding this process, please bring them up at the developer meeting.

3. Create a Feature Proposal
===============================
If the feature request is deemed feasible, the development team will create a feature proposal that extensively describes the proposed feature. This step will be performed by the development team and will consist of the artifacts mentioned in the model "Create Feature Proposal" below:

.. figure:: ./aremtis-feature-proposal-flow.png
    :align: center
    :alt: Create a Feature Proposal - Activity Diagram
    :figclass: align-center

> Step 1: Amend Feature Proposal Template to Feature Request on GitHub

Reviewer 5: ```suggestion
Step 1: Append Feature Proposal Template to Feature Request on GitHub
```

---------------------------------------------------------------------
The Feature Proposal Template outlines the structure of the feature proposal and provides a guideline for the development team to follow.

We have written a Github Action that automatically attaches the feature proposal template to the issue description of an issue when someone is assigned.
As a developer who is responsible for the implementation of a feature request, you do not have to do any than register as an assignee in the issue. 

.. note::
      The GitHub Action will be skipped for issues that are labeled with "bug".

.. literalinclude:: ../../../.github/ISSUE_TEMPLATE/feature-proposal--developer-.md
    :caption: Artemis/.github/ISSUE_TEMPLATE/feature-proposal--developer-.md

>     :language: markdown

Author: Again, the process documented here reflects what was discussed and agreed upon in multiple developer meetings. This current documentation simply implements the previously established process and is not set in stone. 
If you have concerns regarding this process, please bring them up at the developer meeting.

What is UICompanion?
--------------------
UICompanion is a combination of a GitHub bot and a Figma plugin designed to support developers working on the Artemis project through the UI/UX design step. This tool automates several aspects of the design process, making it easier for developers to create and iterate on UI designs.

How to Use UICompanion
----------------------
When an issue is labeled with "UI" on GitHub, UICompanion automatically attaches a comment to the issue prompting the developer to start the UI design process in Figma. Here's how to proceed:

1. **Initiate Design Process**: Upon receiving the UICompanion comment, click on the provided Figma Project Link to access our Figma project.
2. **Follow Guidance**: Inside Figma, use the UICompanion Figma app for step-by-step guidance through the design process. This includes creating prototypes and seeking approval from designated Artemis maintainers.
3. **Automate Tasks**: UICompanion helps automate tasks such as adding the necessary UI components from the Artemis Design System, ensuring that designs remain consistent with the project's UI guidelines.

> For more details and to get started with UICompanion, visit the `UICompanion Repository on GitHub <https://github.com/ls1intum/UICompanion>`_.

Author: UICompanion is not yet in production. Will add the section back in a follow-up PR once it is ready.

.. note::
      Make sure to follow the `Artemis Code and Design Guidelines <https://docs.artemis.cit.tum.de/dev/guidelines.html>`_.

5. Create a Pull Request
========================
After the feature implementation is complete, the developer is required to create a pull request for integrating the feature into the develop branch. 
The subsequent sections provide guidance on the naming conventions and outline the necessary steps for creating and merging a pull request.

Naming Conventions for GitHub Pull Requests
-------------------------------------------

1. The first term is a main feature of Artemis and should use code highlighting, e.g.  “``Programming exercises``:”.

    1. Possible feature tags are: ``Programming exercises``, ``Quiz exercises``, ``Modeling exercises``, ``Text exercises``, ``File upload exercises``, ``Exam mode``,
       ``Grading``, ``Assessment``, ``Communication``, ``Notifications``, ``Team exercises``, ``Lectures``, ``Plagiarism checks``, ``Learning analytics``,

>        ``Adaptive learning``, ``Learning path``, ``Tutorial groups``, ``Iris``.

Reviewer 5: Make sure this list is complete. It is missing at least "Integrated code lifecyle"

>     2. If the change is not visible to end users, or it is a pure development or test improvement, we use the term “``Development``:”.

Reviewer 5: ```suggestion
    2. If the change is not visible to end users, or it is a pure development or test improvement, we use the term ``Development``.
```

>        ``Adaptive learning``, ``Learning path``, ``Tutorial groups``, ``Iris``, ``Scalability``, ``Usability``, ``Performance``, ``Intrastructure``, ``Mobile Apps``.

Author: The feature tags are now capitalized, adhering to the standard English practice for titling. The tags for Usability, Performance, and Infrastructure are directly taken from the Artemis README, as confirmed by on of the Artemis maintainers.

    1. The text should be short, non-capitalized (except the first word) and should include the most important keywords. Do not repeat the feature if it is possible.
    2. We generally distinguish between bugfixes (the verb “Fix”) and improvements (all kinds of verbs) in the release notes. This should be immediately clear from the title.
    3. Good examples:

        - “Allow instructors to delete submissions in the participation detail view”
        - “Fix an issue when clicking on the start exercise button”
        - “Add the possibility for instructors to define submission policies”

Steps to Create and Merge a Pull Request
----------------------------------------

0. Preconditions (For Developers Only)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
* Limit yourself to one functionality per pull request.
* Split up your task in multiple branches & pull requests if necessary.
* `Commit Early, Commit Often, Perfect Later, Publish Once. <https://speakerdeck.com/lemiorhan/10-git-anti-patterns-you-should-be-aware-of>`_

> 1. Begin Implementation (For Developers Only)

Reviewer 5: (Not for this PR but for future reference)
We should also discuss in the meeting if opening a draft PR at the start of development is in-line with how we handle PRs in the future. (See also [meeting notes](https://confluence.ase.in.tum.de/display/ArTEMiS/Artemis+Developer+Meeting+-+2024-04-22) from the last dev meeting)
![image](https://github.com/ls1intum/Artemis/assets/118574504/d640e0bd-ce83-404a-85db-6797ddb93095)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
* `Open a draft pull request. <https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request>`_ This allows for code related questions and discussions.

2. Complete Implementation (For Developers Only)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
* Make sure all steps in the `Checklist <https://github.com/ls1intum/Artemis/blob/develop/.github/PULL_REQUEST_TEMPLATE.md>`_ are completed.
* Add or update the "Steps for Testing" in the description of your pull request.
* Make sure that the changes in the pull request are only the ones necessary.

> * Mark the pull request as `ready for review. <https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request>`_

Author: The process documented here reflects what was discussed and agreed upon in multiple developer meetings. For additional guidelines on writing motivations for PRs without a specific feature proposal or bug report, I recommend addressing this in a follow-up PR. This current documentation simply implements the previously established process and is not set in stone.

    * conform to the code style guide.
    * are easily understandable.
    * contain thorough documentation where applicable.
    * maintain reasonable performance (e.g. no excessive/unnecessary database queries or HTTP calls).
* Submit your comments and status (👍 Approve or 👎 Request Changes) using GitHub.
    * Explain what you did (test, review code) and on which test server in the review comment.

4. Address Review Comments
^^^^^^^^^^^^^^^^^^^^^^^^^^
**For Developers:**

* Use the pull request to discuss comments or ask questions.
* Update your code where necessary.
* Revert to draft if the changes will take a while during which review is not needed/possible.
* Set back to ready for review afterwards.
* Notify the reviewer(s) once your revised version is ready for the next review.
* Comment on "inline comments" (e.g. "Done").

**For Reviewers:**

* Respond to questions raised by the developer.
* Mark conversations as resolved if the change is sufficient.

.. note::
      Repeat steps 3 & 4 until the pull request is approved by all designated reviewers. It's only ready to merge when it has approvals from at least four reviewers.

5. Merge Changes
^^^^^^^^^^^^^^^^

> A project maintainer merges your changes into the ``develop`` branch.

Author: The additional steps regarding feature-specific and Artemis Maintainers' reviews have now been incorporated into section 4 of the documentation.
